src/main/java/hudson/remoting/FileSystemJarCache.java

```diff
@@ -56,6 +56,14 @@ protected URL lookInCache(Channel channel, long sum1, long sum2) throws IOExcept
     protected URL retrieve(Channel channel, long sum1, long sum2) throws IOException, InterruptedException {
         File target = map(sum1, sum2);
         File parent = target.getParentFile();
+
+        if (target.exists()) {
+            // Assume its already been fetched correctly before. ie. We are not going to validate
+            // the checksum.
+            LOGGER.fine(String.format("Jar file already exists: %16X%16X", sum1, sum2));
+            return target.toURI().toURL();
```

Comment:
Under the assumption that this is the rare case and so not performance-sensitive, yes it may be worth checking here. I do not know offhand if this is called only as a race condition, or if this is a normal cache hit which must be fast.